saq/queue.py

```diff
@@ -743,9 +744,10 @@ async def _daemon(self) -> None:
                 logger.exception("Failed to consume message")
 
     async def start(self) -> None:
-        if not self._daemon_task:
-            await self.pubsub.psubscribe(f"{self.prefix}*")
-            self._daemon_task = asyncio.create_task(self._daemon())
+        async with self._daemon_creation_lock:
```

Comment:
nit:

How about this?:
```python
if not self._daemon_task:
    async with self._lock:
        if not self._daemon_task:
            ...
```

Reasoning: its much quicker to check a boolean than it is to acquire lock. We expect this lock to only be needed once, so we're sacrificing some elegance for speed.

Reply:
I've been thinking about it

Leads to the following problem

```
test
test
test
test
test
test
test
test
test
test
Failed to consume message
Traceback (most recent call last):
  File "/home/evgeniy/rq/saq/saq/queue.py", line 737, in _daemon
    message = await self.pubsub.get_message(timeout=None)  # type: ignore
  File "/home/evgeniy/rq/venv/lib/python3.10/site-packages/redis/asyncio/client.py", line 1042, in get_message
    response = await self.parse_response(block=(timeout is None), timeout=timeout)
  File "/home/evgeniy/rq/venv/lib/python3.10/site-packages/redis/asyncio/client.py", line 915, in parse_response
    response = await self._execute(
  File "/home/evgeniy/rq/venv/lib/python3.10/site-packages/redis/asyncio/client.py", line 895, in _execute
    return await conn.retry.call_with_retry(
  File "/home/evgeniy/rq/venv/lib/python3.10/site-packages/redis/asyncio/retry.py", line 62, in call_with_retry
    await fail(error)
  File "/home/evgeniy/rq/venv/lib/python3.10/site-packages/redis/asyncio/client.py", line 884, in _disconnect_raise_connect
    raise error
  File "/home/evgeniy/rq/venv/lib/python3.10/site-packages/redis/asyncio/retry.py", line 59, in call_with_retry
    return await do()
  File "/home/evgeniy/rq/venv/lib/python3.10/site-packages/redis/asyncio/connection.py", line 541, in read_response
    response = await self._parser.read_response(
  File "/home/evgeniy/rq/venv/lib/python3.10/site-packages/redis/_parsers/hiredis.py", line 206, in read_response
    await self.read_from_socket()
  File "/home/evgeniy/rq/venv/lib/python3.10/site-packages/redis/_parsers/hiredis.py", line 186, in read_from_socket
    raise ConnectionError(SERVER_CLOSED_CONNECTION_ERROR) from None
redis.exceptions.ConnectionError: Connection closed by server.
```
I assumed that it happens because I need to add same lock to close() as well.

But that doesn't solve the problem

```
    async def close(self) -> None:
        await self.pubsub.punsubscribe()
        if self._daemon_task:
            async with self._daemon_lifespan_lock:
                self._daemon_task.cancel()
                await self._daemon_task
                self._daemon_task = None
                self._subscriptions.clear()
                self._queues.clear()
```


any guesses on that?

Reply:
Stepping AFK for a bit, will take a look in a few hours

Reply:
Oops, I missed that you had the if not self._daemon_task: check twice. I only added it once, not realizing it doesn't fully solve the issue. With two checks, it looks good now.